feat(cdk): aspect framework and tag implementation

moofish32 opened the pull request:
A generalized aspect framework is added. Aspects can be used to affect the construct tree without modifying the class hierarchy. This framework is designed to be generic for future use cases. Tagging is the first implementation.

Tagging has been reimplemented to leverage the aspect framework and simplify the original tag design. Tag Manager still exists, but is no longer intended for use by L2 construct authors. There are two new classes `cdk.Tag` and `cdk.RemoveTag`. As new resources are added tag support will be automatic as long as they implement one of the existing tag specifications. All L2 constructs have removed the TagManager.

Code generation has been modified to add tag support to any CloudFormation Resource with a matching specification, by creating a Tag Manager for that resource as a `tags` attribute. The schema code now includes the ability to detect 3 forms of tags which include the current CloudFormation Resources.

BREAKING CHANGE: if you are using TagManager the API for this object has completely changed. You should no longer use TagManager directly, but instead replace this with Tag Aspects. `cdk.Tag` has been renamed to `cdk.CfnTag` to enable `cdk.Tag` to be the Tag Aspect.

Fixes #1136
Fixes #1497 

----

### Pull Request Checklist

* [x] Testing
  - Unit test added
  - __CLI change?:__ manually run [integration tests](https://github.com/awslabs/aws-cdk/blob/master/packages/aws-cdk/integ-tests/test.sh) and paste output as a PR comment
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

moofish32 commented on packages/@aws-cdk/cdk/lib/cloudformation/resource.ts:
```diff
@@ -48,6 +49,15 @@ export class Resource extends Referenceable {
+   *
+   * This object is rendered via a call to "renderProperties(this.properties)".
+   */
+  public readonly properties: any;
```
Yep, I'm going to have refactor `renderProperties` to make a super call in the generated code. I'll propose something in the next refactor.

moofish32 commented on packages/@aws-cdk/cdk/lib/cloudformation/resource.ts:
```diff
@@ -239,6 +242,47 @@ export class Resource extends Referenceable {
+/**
+ * Handles converting TagManager tags to AutoScalingGroupTags
+ */
+export class AutoScalingGroupTagFormatter implements ITagFormatter {
```
Yes they are already in the aspect this is just not cleaned up yet do to my rush on the draft.

rix0rrr replied:
Yes, these are redundant now, right?

moofish32 commented on packages/@aws-cdk/cdk/lib/core/construct.ts:
```diff
@@ -299,6 +304,22 @@ export class Construct {
+   * TODO docs
+   * can this be private?
+   */
+  public visit(): void {
```
agree - even in Wikipedia they call apply `Accept` and `Visit` does the work. I'm open to name changes.

rix0rrr replied:
My naming suggestion here is semantic meaning from the perspective of the user, not mechanic meaning from the perspective of the implementor.

For the user, it looks like they are *applying* a tag to a construct, and the word *apply* is nice and generic enough that we can apply it (hah) to other types of aspects as well.

A mechanically correct alternative name would be `attachAspect()`, but it doesn't read as nicely.

Compare:

```ts
// Customer first (what it does)
construct.apply(new cdk.Tag('key', 'value'));

// Implementor first (how it works)
construct.attachAspect(new cdk.TagAspect('key', 'value'));
```

At least, im(ns)ho :)

rix0rrr replied:
Also, I wouldn't call this particular function `visit()`, I would call this something like `invokeAspects()`.

And yes, it shouldn't be public, that will only serve to confuse people. It's not too bad, it should be invoked from `App.run()` which is itself a construct so it can access `protected` members.

moofish32 commented on packages/@aws-cdk/cdk/lib/core/construct.ts:
```diff
@@ -417,6 +438,14 @@ export class Root extends Construct {
+ * Aspects that can be applied to Constructs.
+ */
+export interface IAspect {
+  readonly type: string;
```
This type was included so we could filter on aspect types. For example, TagAspects check to see if there is another TagAspect that takes precedence. It's hard because parents and children can apply tags out of order and I had to rely on more than array order. There are other designs that could be done to make this work. For example, we could make the TagAspect collect all others TagAspects into an array and write a comparison function to always sort priority.

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts:
```diff
@@ -0,0 +1,149 @@
+    }
+    this.visitedBy[node.uniqueId] = true;
+    const propagate = this.propagate > 0 ? this.propagate - 1 : this.propagate;
+    this.parent = this.parent || node;
```
If you need this, it should also become part of the framework.

f.ex:

```ts
public visit(currentNode: Construct, applicationNode: Construct) {
```

fwiw I think it makes sense.

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts:
```diff
@@ -0,0 +1,149 @@
+    if (propagate !== 0) {
+      for (const child of node.children) {
+        this.propagate = propagate;
+        child.apply(this);
```
Agreed, I think the framework should recursively apply.

rix0rrr replied:
Also, in my head the Tags where physically attached only to the construct the user `apply()`d them to, and only during `visit()`ation would they also `visit()` downward in the tree. Because... (:asterisk:, see below)

moofish32 replied:
I agree with all of these comments, but some of the comments conflict.  If `visit()` recurses then I think I need some form of a collector on the resources because mutating `properties` we've also agreed is not acceptable. That brings back something like `TagManager`. If I recurse in `apply` at the `Construct` level then our simplified ordering story *️⃣ below I think becomes difficult again. Is there something I'm missing that avoids these draw backs?

I leaning towards `visit` moving the tag down the tree and combining `TagType` with a simplified `TagManager` ideally named just `Tags`.

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts:
```diff
@@ -0,0 +1,149 @@
+  }
+}
+
+export class TagAspect extends TagBaseAspect {
```
I know this is going to be contentious but I think I'd prefer this to *not* be suffixed with `Aspect`.

```
construct.apply(new cdk.Tag('key', 'value'));
```

Reads better to me.

moofish32 replied:
I'll agree this reads better and I originally I started here. For this draft I didn't want to rename the current `cdk.cloudformation.Tag` class that we have. I did some looking and the specific class is mostly used in generated code. Would we prefer that we rename this to `CfnTag`. I realize `Cfn<resource>` now has special meaning and I think this is similar, but not identical. I also felt like the fact `apply` (or another name we choose) will already set the mental context of the developer that this is an `Aspect` and suffix is not necessary.

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts:
```diff
@@ -0,0 +1,149 @@
+  }
+}
+
+function setTag(tag: TagBaseAspect, resource: Resource) {
```
Exactly this function is why I wanted to keep `TagManager`. Now, TagManager doesn't need to do complicated work anymore, but the type-specific information about how a tag should be rendered can be kept there.

On the other hand, if we're sure `TagType` does not need to be extensible... I guess okay.

I would prefer for `setTag()` to either take the actual tag key and value if you're going to keep it a free function, or become a member function of `TagBaseAspect` and use `this`.

Also, I would prefer for the second call to `setTag()` to overwrite the first, if present.

moofish32 replied:
I think this moves to base, but let me see if removing propagate changes my opinion.

> Also, I would prefer for the second call to setTag() to overwrite the first, if present.
You mean as opposed to my use of delete then create? Find the index and just replace if there is a second call, is what you would prefer?

rix0rrr replied:
Yes.

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts:
```diff
@@ -0,0 +1,149 @@
+    }
+    case TagType.AutoScalingGroup: {
+      resource.properties.tags = resource.properties.tags || [];
+      resource.properties.tags.push({ key: tag.key, value: tag.value, propagateAtLaunch: tag.propagate !== 0});
```
I'm totally fine with `propagateAtLaunch` being a property of `Tag` (maybe called `applyToLaunchedInstances`), which only applies in case the tag is applied to AutoScalingGroups.

This mechanism of "having enough propagates left once the tag gets applied" seems hard to understand.

rix0rrr commented on packages/@aws-cdk/cdk/lib/core/tag-manager.ts:
```diff
@@ -85,6 +86,18 @@ export interface RemoveProps {
+ * specialized CloudFormation Resources (e.g. AutoScalingGroups, Serverless and
+ * DAX)
+ */
+export interface ITagFormatter {
```
Deprecated, right?

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts:
```diff
@@ -0,0 +1,149 @@
+
+  protected abstract applyTag(resource: Resource): void;
+
+  private hasPrecedence(aspects: IAspect[]): boolean {
```
:asterisk:, if we apply in order top-down in the tree, tags applied lower in the tree will naturally overwrite the tags applied higher in the tree, without having to write this function. Let's assume the following tree:

```
┌────────────────┐                    
│                │                    
│     Parent     │    Tag(key, value1)
│                │                    
└────────────────┘                    
         │                            
         ▼                            
┌────────────────┐                    
│                │                    
│     Child      │   Tag(key, value2) 
│                │                    
└────────────────┘                    
         │                            
         ▼                            
┌────────────────┐                    
│                │                    
│TaggableResource│                    
│                │                    
└────────────────┘                            
```

This would resolve in the following way:

* The tag with `value1` would apply first. It would try to apply to `Parent` (doesn't do anything), then it would try to apply to `Child` (doesn't do anything), then it would apply to `TaggableResource` (sets `key: value1`).
* Next, the tag with `value2` would apply. It would try to apply to `Child` (doesn't do anything), then it would apply to `TaggableResource`. The value for tag `key` would be overwritten with `value2`.

Presto, the more specific tag has taken precedence without any extra work on the aspect implementor's part.

rix0rrr commented:
> Neither of you commented on include/exclude API [...] I believe the include/exclude feature handles my selective application needs 

I would prefer a more generic selector mechanism that also works for other aspects. However I recognize that we don't have a great solution for this right now, and at least it doesn't harm to have a mechanism inside the Tag aspect system as well; we can always refactor later. So I'm okay with keeping it in for now.

rix0rrr commented:
> And yes, it shouldn't be public, that will only serve to confuse people. It's not too bad, it should be invoked from App.run() which is itself a construct so it can access protected members.

That may get you into trouble for unit tests, btw. But in those cases we can always do the pragmatic thing of casting to `any` to get rid of any type checking :) or declaring a public method on `App` which does the "pre-synthesis" work of invoking aspects, and rewriting tests to use `App` explicitly.

moofish32 commented:
Ok I'm at test passing, unit test complete, and jsdocs complete. 

I added a `TestStack` to `assert` that will handle the invoking of aspects. Plus it appeared we were creating this entity in a lot of places. I also cleaned up some left over `tslint` flags as I was searching for my own. I had to update the ECS tests because it appears the tag from the ASG wasn't reaching the child Lambdas. I think this was a bug, however, if that was intentional we can revert.

moofish32 commented on packages/@aws-cdk/assert/lib/test-stack.ts:
```diff
@@ -0,0 +1,7 @@
+import cdk = require('@aws-cdk/cdk');
+
+export class TestStack extends cdk.Stack {
```
1. Is this a good idea to put in a reusable place?
1. I started to find all the places we *should* use this, but I missed a few and now I'm questioning if this is too much to include here?

moofish32 replied:
I think this class is no longer needed but the test clean up seems like it was a good idea?

eladb commented on packages/@aws-cdk/assert/lib/test-stack.ts:
```diff
@@ -0,0 +1,7 @@
+
+export class TestStack extends cdk.Stack {
+  public testInvokeAspects() {
+    this.invokeAspects();
```
Should we move  invokeAspects behind the all new “prepareTree”?

rix0rrr replied:
Yes I think that makes sense.

moofish32 replied:
Just for clarity, I am doing:

1. remove `invokeAspects()` from `app.ts`
1. putting `invokeAspects()` inside of `prepareTree()`
1. removing the `TestStack` because `prepareTree()` is public and called during `expect`

Is there any concern about the conditional wrapped around the block that calls `prepareTree()`?

```ts
    if (!skipValidation) { // any concern here - I haven't found any but just double checking
      // Do a prepare-and-validate run over the given stack
      stack.node.prepareTree();

      const errors = stack.node.validateTree();
      if (errors.length > 0) {
        throw new Error(`Stack validation failed:\n${errors.map(e => `${e.message} at: ${e.source.node.scope}`).join('\n')}`);
      }
    }

// this will be removed
    if (isTestStack(stack)) {
      stack.testInvokeAspects();
    }
// end of block to be removed
```
A simple thumbs up to verify is fine.

eladb replied:
Lgtm; Curious why prepare tree is skipped if validation is skipped but that’s for a different issue (raise one?).

eladb commented on packages/@aws-cdk/cdk/lib/aspects/aspect.ts:
```diff
@@ -0,0 +1,54 @@
+   * Node will only be visited once.
+   */
+  public visit(construct: IConstruct): void {
+    if (this.visitedBy[construct.node.uniqueId] === true) {
```
I’d rename this to “visitTree” and “visitAction” to “visit”

moofish32 replied:
A little clarification. Does this also mean change:

```ts
/**
 * Represents an Aspect
 */
export interface IAspect {
  /**
   * The type of Aspect
   */
  readonly type: string;

  /**
   * All aspects can visit by IConstructs
   */
  visit(node: IConstruct): void;
}
```

to 

```ts
/**
 * Represents an Aspect
 */
export interface IAspect {
  /**
   * The type of Aspect
   */
  readonly type: string;

  /**
   * All aspects can visit the construct tree starting from this construct
   */
  visitTree(node: IConstruct): void;
}
```

eladb replied:
I wouldn’t say so. I think aspects are only expected to deal with their own visit. The tree visitation is handled by the framework, no?

eladb commented on packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts:
```diff
@@ -0,0 +1,120 @@
+   * An array of Resource Types that will receive this tag
+   */
+  include?: string[];
+
```
What are the defaults?

moofish32 replied:
I'll add to the docs

eladb commented on packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts:
```diff
@@ -0,0 +1,120 @@
+  /**
+   * Test if the construct is a CloudFormation Resource
+   */
+  public static isResource(resource: any): resource is Resource {
```
Move this to the Resource class (Resource.isResource is the convention)

moofish32 replied:
will fix

eladb commented on packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts:
```diff
@@ -0,0 +1,120 @@
+  /**
+   * The string value of the tag
+   */
+  public readonly value?: string;
```
Value is only needed for Tag and not for RemoveTag so why is it in TagBase?

moofish32 replied:
yep - a mistake for sure, will fix

eladb commented on packages/@aws-cdk/cdk/lib/cloudformation/resource.ts:
```diff
@@ -48,6 +62,13 @@ export class Resource extends Referenceable {
+   *
+   * This object is rendered via a call to "renderProperties(this.properties)".
+   */
+  protected readonly properties: any;
```
Why do we need to expose this?

moofish32 replied:
https://github.com/awslabs/aws-cdk/blob/master/packages/@aws-cdk/cdk/lib/cloudformation/resource.ts#L68 I was reverting back to master. Isn't it `protected` there already?

eladb commented on packages/@aws-cdk/cdk/lib/cloudformation/resource.ts:
```diff
@@ -239,6 +257,13 @@ export class Resource extends Referenceable {
   }
 }
 
+export enum TagType {
```
Indentation

moofish32 commented on packages/@aws-cdk/cdk/lib/aspects/aspect.ts:
```diff
@@ -0,0 +1,52 @@
+  /**
+   * All aspects can visit by IConstructs
+   */
+  visitTree(node: IConstruct): void;
```
@eladb let's take a good look here to make sure I implemented what you want. I *think* I want a method that calls the action an Aspect author creates. I want this method so I can manage the checking of the visit status. If that checking was in `visit()` then the author would have to remember to implement it. (this is implemented in the PR currently)

The other option is that we could move the management of which aspect were visited by which node, to the node itself. This is going to add more to ConstructNode but then it's clearly hooked on to the framework.

eladb replied:
+1 for moving to the framework

moofish32 replied:
Ok, let me make that move. I don't think it will be very complicated.

moofish32 commented on packages/@aws-cdk/cdk/lib/core/construct.ts:
```diff
@@ -447,6 +458,24 @@ export class ConstructNode {
+      return aspect.visitType === AspectVisitType.Multiple ||
+        !this.invokedAspects.includes(aspect);
+    });
+  }
```
The functions above here could be extracted into a class for composition. They were not very complicated so I left them here. If we went to a class we could likely also grab the two arrays I've added to construct and move those into the class. @rix0rrr @eladb -- let me know if you have a strong opinion here.

moofish32 commented on packages/@aws-cdk/cdk/lib/aspects/aspect.ts:
```diff
@@ -0,0 +1,27 @@
+  /**
+   * Aspects invocation pattern
+   */
+  readonly visitType: AspectVisitType;
```
First, with the move of aspect visit only once to the framework (`ConstructNode`) the need for a base `Aspect` is gone. Speak up if you disagree.

With all the simplification (thanks team for the comments), I no longer need a string type here. @eladb and I were talking about multiple visit pattern and I added it here. Looking at `prepareTree()` and where this sits I don't think the multiples make a lot of sense.  Two questions for your feedback:
1. Should I bring back the string type?
1. Should we keep `AspectVisitType`?

rix0rrr replied:
It's a little confusing to me. What's the point of the multiple visit type? When will it be useful?

moofish32 replied:
ready to remove -- see comments on the README section.

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,292 @@
+## Aspects
```
Thanks for writing the README, but I feel this doc is focusing on implementation perspective a little too much over customer perspective. I'm going to add some more comments over the course of it.

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,292 @@
+
+### Tag and Remove Tag Aspect
+
+Tag support is the first use of Aspects. The goal was to enable the ability to
```
Not sure that people care whether or not Tags use Aspects, I think they only care about how tagging works.

eladb replied:
I think aspects are an important (albeit advanced concept) which @Doug-AWS  should definitely add to our developer guide.

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
````diff
@@ -0,0 +1,292 @@
+// snip //
+```
+
+If you run the above modification you will notice that the Subnets, Route
````
You are correct in that the design I sent you towards has this issue. But rather than have a giant documentation section that explains the problem and gives you a horrible way to work around it, shouldn't we resolve the issue instead?

A simple solution would be to add an (optional) `priority` field to every Tag, and have it so that by default `RemoveTag.priority > Tag.priority` (so that `RemoveTag` takes precedence over `Tag` applied lower). We'll have to tombstone tags to make this work, but that doesn't seem too hard.

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,292 @@
+
+Tag Aspects support three options: include, exclude and applyToLaunchInstances.
+
+##### applyToLaunchInstances
```
`launchEDinstances` seems to read better to me?

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,292 @@
+
+1. Single visit aspects will ensure the aspect visit function is only invoked
+   one time.
+1. Multiple visit aspects will invoke the visit function as many times as
```
I don't understand what this means though.

moofish32 replied:
I am ready to rip out the this concept. @rix0rrr @eladb -- looking for agreement from you guys.

Also this means the interface for aspect is simplified to:

```ts
export interface IAspect {
  /**
   * All aspects can visit a IConstruct
   */
  visit(node: IConstruct): void;
}
```

Are you good with that too?

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,292 @@
+visits an AutoScalingGroup resource the `PropagateAtLaunch` property is set to
+true. If false the property is set accordingly.
+
+##### include
```
Maybe call `includeResourceTypes` then?

rix0rrr commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,292 @@
+resource types in the array. By default the array is empty and an empty array is
+interpreted as apply to any resource type.
+
+##### exclude 
```
And similar.

moofish32 commented on packages/@aws-cdk/aws-kms/test/test.key.ts:
```diff
@@ -138,97 +137,95 @@ export = {
-    expect(app.synthesizeStack(stack.name)).to(exactlyMatchTemplate({
+    key.apply(new Tag('tag1', 'value1'));
+    key.apply(new Tag('tag2', 'value2'));
+    key.apply(new Tag('tag3', ''));
```
I had the same thought and based on this: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-resource-tags.html it seems like `1` is the min length.  I'll turn this into a raise on empty string and fix this test.

moofish32 replied:
I'm ready to turn this into a raise. What do you all think?

moofish32 commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,292 @@
+
+| Construct Path | Tag Key | Tag Value |
+| ----------|:---------|:-----|
+|MarketingSystem/MarketingVpc|CostCenter|Platform|
```
shortened this -- but more importantly added an additional section on tagging. Let me know what you think?

moofish32 commented:
@rix0rrr or @eladb -- I've added some comments to our threads above. I've also added to the README in aspects. I don't think that is the final resting place for the docs, but the content is my best effort. If you want this moved or if somebody is going to roll this in somewhere else just let me know. Once we are aligned on the feature, I'll rebase and clean up the commit comment just before we merge (of course fixing the conflicts above).

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+## Tagging with the CDK
```
This is not the correct place for this README as it won't be surfaced in our docs this way. Can you move this to the @aws-cdk/cdk README? I know it's pretty empty right now, but basically that's where it should go. Simply append this README to that one.

Doug-AWS replied:
Sure. The major change from this README to the guide will be that the guide does not discuss implementation. The purpose of a guide is to answer the questions:

- What is this feature?
- Why should I care?
- How do I use it?

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+## Tagging with the CDK
+
+Tags are implemented using `Aspects` for details on the implementation see the
```
Can we dedicate a separate section to aspects before we introduce tags?

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+Tags are implemented using `Aspects` for details on the implementation see the
+detailed sections below.
+
+### Tag Quickstart
```
I would just dive right into it without the "tag quickstart" title

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
````diff
@@ -0,0 +1,345 @@
+tag:
+
+```ts
+import cdk = require('@aws-cdk/cdk');
````
Preferably make this more idiomatic:

```ts
class MyStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.apply(new cdk.Tag('StackType', 'TheBest'));
  }
}

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
````diff
@@ -0,0 +1,345 @@
+// any resources added that support tags will get them
+```
+
+### Advanced Tagging
````
Redundant heading

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
````diff
@@ -0,0 +1,345 @@
+Types are affected you can include or exclude those in properties.
+
+```ts
+import cdk = require('@aws-cdk/cdk');
````
I don't think you need the whole program every time. Only the tagging related snippets will do:

```ts
this.apply(new cdk.Tag(..., {
  includeResourceTypes: [ ec2.CfnVpc.resourceType ]
});
```

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+const theBestStack = new cdk.Stack(app, 'MarketingSystem');
+theBestStack.apply(new cdk.Tag('StackType', 'TheBest', {
+  includeResourceTypes: [ec2.CfnVpc.resourceType]
+  }));
```
indentation

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
````diff
@@ -0,0 +1,345 @@
+
+```ts
+import cdk = require('@aws-cdk/cdk');
+
````
idiomatic..

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+const app = new cdk.App();
+const theBestStack = new cdk.Stack(app, 'MarketingSystem');
+theBestStack.apply(new cdk.Tag('StackType', 'TheBest', {
+  includeResourceTypes: [ec2.CfnVpc.resourceType]
```
Not sure I like that we need to use L1s to reference resource types... raise an issue (we can add to L2s)?

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
````diff
@@ -0,0 +1,345 @@
+
+```
+
+#### Removing Tags
````
Change to H3s

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+const app = new cdk.App();
+const theBestStack = new cdk.Stack(app, 'MarketingSystem');
+
+theBestStack.apply(new cdk.Tag('StackType', 'TheBest', {prioirty: 1}));
```
Leave only tagging code

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+
+You can use priority to control tagging regardless of the hierarchy of the tree.
+
+## Aspects
```
Move before tags

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+## Aspects
+
+Aspects are a mechanism to extend the CDK without having to directly impact the
+class hierarchy directly. We have implemented Aspects using the [Visitor
```
"directly" twice in the same sentence.

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+class hierarchy directly. We have implemented Aspects using the [Visitor
+Pattern](https://en.wikipedia.org/wiki/Visitor_pattern).
+
+### Aspects in the CDK
```
Redundant heading

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+ */
+export interface IAspect {
+  /**
+   * All aspects can visit an IConstruct
```
Not sure what this description means unneeded

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
````diff
@@ -0,0 +1,345 @@
+}
+```
+
+Aspects can be applied to any `Construct`. During the tree
````
no need to quote "Construct", just use it as a word: "construct"

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+time. Aspects are invoked in the order they were added to the `Construct`,
+starting at the `App` (root of the tree) and progressing in order to the leaf
+nodes (most commonly the CloudFormation Resource). Aspect authors will implement 
+the `visit(IConstruct)` function can inspect the `Construct` for specific characteristics.
```
@Doug-AWS can you help with editing here?

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+the `visit(IConstruct)` function can inspect the `Construct` for specific characteristics.
+Such as, is this construct a CloudFormation Resource?
+
+### Tag and Remove Tag Aspect
```
This should go to the Tagging section under an `API` heading or something like that

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+exclude a CloudFormation Resource Type. See the TagAspectProps interface for
+more details. 
+
+#### Tag Example with ECS
```
This should probably go to `examples` and only referenced here

Doug-AWS replied:
Let's keep it simple. We now support tags. You can:

- Add (apply?) a tag, either to specific resources or all but specific resources
- Remove a tag, again either from specific resources or all but specific resources

Tags are inherited, based on the scope. If you tag resource A, and A contains resource B, resource B inherits the tag.

And then code examples, followed by some API-ish text and a link to the reference.

eladb commented on packages/@aws-cdk/cdk/lib/aspects/README.md:
```diff
@@ -0,0 +1,345 @@
+// every resource starts with Marketing
+app.apply(new cdk.Tag(COST_CENTER_KEY, 'Marketing'));
+
+const marketingStack = new cdk.Stack(app, 'MarketingSystem');
```
Please make this code more idiomatic (extend `Stack` instead of a flat app)

eladb commented:
@moofish32 we are getting there. your fault for picking up a core feature... our bar is very high at that layer... 

can you take some time to rework the PR description to make sure it's what you intend to use for the commit message?

Doug-AWS commented:
I'll try to create my version of README.md before Thursday.

moofish32 commented:
@Doug-AWS - I tried to incorporate some of the feedback. Hopefully I'm helping not colliding with you. @eladb I took most of your feedback and attempted to make this flow a little better. I still owe you a commit message update and I'll get that done shortly.

moofish32 commented:
@eladb updated commit message, I'll fix the line returns and resolve conflicts if we think I've covered it.

moofish32 commented:
I might need a hint on this latest jsii error?

eladb commented on examples/cdk-examples-typescript/hello-cdk-ecs-tags/index.ts:
```diff
@@ -0,0 +1,53 @@
+
+const COST_CENTER_KEY = 'CostCenter';
+
+class MarketingFargate extends cdk.Stack {
```
Maybe call this `MarketingStack` or `MarketingDepartmentStack`

eladb commented on packages/@aws-cdk/cdk/README.md:
```diff
@@ -3,3 +3,149 @@
+An aspect in the CDK is defined by this [interface](lib/aspects/aspect.ts)
+
+Aspects can be applied to any construct. During the tree
+preparation phase the aspect will visit each construct in the tree one
```
change from `preparation` to `"prepare"` (with quotes)

eladb commented on packages/@aws-cdk/cdk/README.md:
```diff
@@ -3,3 +3,149 @@
+
+Aspects can be applied to any construct. During the tree
+preparation phase the aspect will visit each construct in the tree one
+time. Aspects are invoked in the order they were added to the construct,
```
"one time" => "once"

eladb commented on packages/@aws-cdk/cdk/README.md:
```diff
@@ -3,3 +3,149 @@
+preparation phase the aspect will visit each construct in the tree one
+time. Aspects are invoked in the order they were added to the construct,
+starting at the `App` (root of the tree) and progressing in order to the leaf
+nodes (most commonly the CloudFormation Resource). Aspect authors will implement 
```
Add "breadth first"

eladb replied:
Remove the word "will": "aspect authors implement the ..."

eladb commented on packages/@aws-cdk/cdk/README.md:
```diff
@@ -3,3 +3,149 @@
+
+## Tagging
+
+Tags are implemented using Aspects.
```
Use "aspects" as a noun without capitalization. Once you defined it, no need to capitalize everywhere

eladb commented on packages/@aws-cdk/cdk/README.md:
```diff
@@ -3,3 +3,149 @@
+The Tag API supports: 
+
+ * Add (apply) a tag, either to specific resources or all but specific resources
+ * Remove a tag, again either from specific resources or all but specific resources
```
We need to get @Doug-AWS to work on this

eladb commented on packages/@aws-cdk/cdk/README.md:
```diff
@@ -3,3 +3,149 @@
+you tag construct A, and A contains construct B, construct B inherits the tag.
+The Tag API supports: 
+
+ * Add (apply) a tag, either to specific resources or all but specific resources
```
Add the names of the relevant classes for each

eladb commented on packages/@aws-cdk/cdk/README.md:
````diff
@@ -3,3 +3,149 @@
+// any resources added that support tags will get them
+```
+
+The goal was to enable the ability to define tags in one place and have them 
````
Put under a block quote

eladb commented on packages/@aws-cdk/cdk/README.md:
```diff
@@ -3,3 +3,149 @@
+
+In order to enable additional controls a Tags can specifically include or
+exclude a CloudFormation Resource Type, propagate tags for an autoscaling group,
+and use priority to override the default precedence. See the `TagAspectProps` 
```
This should be called `TagProps`, no?

eladb replied:
Also show a quick snippet on how to use these options

moofish32 replied:
What do you think if I just move the `includeResourceType` and exclude to the `TagProps` interface. Then have `TagManager` own all the logic for applying the tag to the resource. Right now it's split because `priority` had to managed at the `TagManager` and include/exclude was done in the `visit` logic. I think eliminating `TagAspectProps` moving all this to `TagProps` is the right choice.

eladb replied:
Yes, makes sense

eladb commented:
Please update the PR title + description to match the commit description you wish us to merge eventually

eladb commented on packages/@aws-cdk/cdk/test/core/test.tag-manager.ts:
```diff
@@ -4,34 +4,34 @@ import { TagManager } from '../../lib/core/tag-manager';
 export = {
   '#setTag() supports setting a tag regardless of Type'(test: Test) {
-    const notTaggable = new TagManager(TagType.NotTaggable);
+    const notTaggable = new TagManager(TagType.NotTaggable, '');
```
put something like "AWS::Resource::Type" instead of empty strings so it's clearer what this string is

moofish32 commented:
@eladb - I think tests are finally passing. Is there anything else?
